keymanapp/keyman
# feat(web): enables mouse interactivity for the predictive banner 🧵

Follows from #5665.

Addresses part of #3963, integrating 15.0's abstraction for OSK event handling into the 12.0 `SuggestionBanner` and `UITouchHandlerBase`.  This may only be one aspect of the issue, but it's a major step in the right direction.  Also includes some (required-for-this) polish that I knew needed to be done anyway for `InputEventEngine`, as it was a bit too specific.

As hinted by the title, the predictive-text banner now accepts mouse-based interaction when on devices without touch enabled.  This will be of particular use for #5723, but it's also a notable step toward supporting predictive text more broadly and openly within KMW.

## User Testing 

### SUITE_REGRESSION - ensure that standard predictive-text banner operations continue as normal

**GROUP_IPHONESE** - test on a Simulator-based iPhone SE, within the Keyman app
**GROUP_GALAXY_FOLD** - test on a Chrome-emulated "Galaxy Fold" (if not available, "Responsive" 280 x 653) on the "Prediction - robust testing" test page

(These are selected because they're among the least-wide [narrowest] test target setups that I know.)

**TEST_LONG_SUGGESTION** - ensure long suggestions are scrollable and may be applied
- type `internat` - `internationally` should appear as the second suggestion.
- As that's a fairly long word, you should be able to horizontally scroll it.  Make sure that's possible and that it works intuitively.
- Ensure you can also tap it to apply the suggestion

### SUITE_MOUSE_USE - tests new functionality

**GROUP_DESKTOP** - use the "Prediction - robust testing" test page _without emulating a mobile device_.  When the OSK is active, you should see something like this:

<img width="636" alt="image" src="https://user-images.githubusercontent.com/25213402/134290173-15764c24-c0a4-42d8-bd24-8bedbe70968e.png">

Yes, it's **ugly**.  We know, but we're not here to test that.  You may also want to resize the OSK to a similar size to facilitate the following test:

**TEST_LONG_SUGGESTION** - ensure long suggestions are scrollable and may be applied.
- type `internat` - `internationally` should appear as the second suggestion.
- As that's a fairly long word, you should be able to horizontally scroll it.  
    - If the full word is visible, your OSK is too wide - resize the OSK to something narrower.
    - Make sure it may be scrolled horizontally and that this works intuitively.
- Shows up easily on iPhone SE
- Ensure you can also tap `internationally` to apply the suggestion.
